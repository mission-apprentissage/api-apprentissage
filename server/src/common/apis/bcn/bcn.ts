import { ReadStream } from "node:fs";

import { internal } from "@hapi/boom";
import { isAxiosError } from "axios";

import { withCause } from "../../errors/withCause";
import getApiClient from "../client";

export type BCN_TABLE =
  | "V_FORMATION_DIPLOME"
  | "N_FORMATION_DIPLOME"
  | "N_FORMATION_DIPLOME_ENQUETE_51"
  | "N_NIVEAU_FORMATION_DIPLOME";

const bcnClient = getApiClient(
  {
    baseURL: "https://infocentre.pleiade.education.fr/bcn",
    timeout: 90_000,
  },
  { cache: false }
);

export async function fetchBcnData(table: BCN_TABLE): Promise<ReadStream> {
  try {
    const response = await bcnClient({
      method: "POST",
      url: "/index.php/export/CSV",
      params: {
        n: table,
        separator: ";",
      },
      responseType: "stream",
    });

    return response.data;
  } catch (error) {
    if (isAxiosError(error)) {
      throw internal("api.bcn: unable to fetchBcnData", { data: error.toJSON() });
    }
    throw withCause(internal("api.bcn: unable to fetchBcnData"), error);
  }
}
