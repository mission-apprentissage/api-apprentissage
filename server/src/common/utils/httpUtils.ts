import axios, { AxiosInstance, AxiosRequestConfig, CreateAxiosDefaults } from "axios";
import http from "http";
import https from "https";
import iconv from "iconv-lite";
import { compose } from "oleoduc";

import logger from "@/common/logger";

// https://github.com/axios/axios/issues/3845#issuecomment-1040819908
class BufferedHttpAgent extends http.Agent {
  highWaterMark: number;

  constructor({ highWaterMark = 16 * 1024, ...rest }) {
    super(rest);

    this.highWaterMark = highWaterMark;
  }
  // @ts-expect-error: TODO
  createConnection(options, callback) {
    // @ts-expect-error
    return super.createConnection({ ...options, highWaterMark: this.highWaterMark }, callback);
  }
}
class BufferedHttpsAgent extends https.Agent {
  highWaterMark: number;

  constructor({ highWaterMark = 16 * 1024, ...rest }) {
    super(rest);
    //see https://github.com/nodejs/node/issues/39092
    this.highWaterMark = highWaterMark;
  }

  // @ts-expect-error: TODO
  createConnection(options, callback) {
    // @ts-expect-error
    return super.createConnection({ ...options, highWaterMark: this.highWaterMark }, callback);
  }
}

type FetchOptions = AxiosRequestConfig & {
  highWaterMark: number;
  raw: boolean;
  encoding: string;
};
async function _fetch(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { method = "GET", data, highWaterMark, ...rest } = options;
  logger.debug(`${method} ${url}...`);

  return client.request({
    url,
    method,
    httpAgent: new BufferedHttpAgent({ highWaterMark }),
    httpsAgent: new BufferedHttpsAgent({ highWaterMark }),
    ...(data ? { data } : {}),
    ...rest,
  });
}

async function fetchStream(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { raw, encoding = "UTF-8", ...rest } = options;
  const response = await _fetch(url, { responseType: "stream", ...rest }, client);

  const stream = compose(response.data, iconv.decodeStream(encoding));

  if (raw) {
    return {
      headers: response.headers,
      stream,
    };
  } else {
    return stream;
  }
}

async function fetchData(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { raw, ...rest } = options;
  const response = await _fetch(url, { responseType: "json", ...rest }, client);
  const data = response.data;

  if (raw) {
    return {
      headers: response.headers,
      data,
    };
  } else {
    return data;
  }
}

async function fetchJson(url: string, options: Partial<FetchOptions> = {}) {
  const response = await _fetch(url, { ...options, responseType: "json" });
  return response.data;
}

function addCsvHeaders(filename: string, encoding: string, res: any) {
  res.setHeader("Content-disposition", `attachment; filename=${filename}`);
  res.setHeader("Content-Type", `text/csv; charset=${encoding}`);
}

const getHttpClient = (options: CreateAxiosDefaults<any> = {}) =>
  axios.create({
    timeout: 15000,
    httpAgent: new http.Agent({ keepAlive: true }),
    httpsAgent: new https.Agent({ keepAlive: true }),
    ...options,
  });

export { addCsvHeaders, fetchJson, fetchStream, fetchData, getHttpClient };
