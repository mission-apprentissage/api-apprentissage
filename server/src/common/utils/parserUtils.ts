import { Options, parse } from "csv-parse";
import { isEmpty, pickBy } from "lodash-es";

export const DEFAULT_DELIMITER = ";";

export function parseCsv(options: Options = {}) {
  return parse({
    trim: true,
    delimiter: DEFAULT_DELIMITER,
    columns: true,
    bom: true,
    relax_column_count: true,
    on_record: (record) => {
      return pickBy(record, (v) => {
        return !isEmpty(v) && v.trim().length;
      });
    },
    ...options,
  });
}
