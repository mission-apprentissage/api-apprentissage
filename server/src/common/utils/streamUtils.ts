import { Transform } from "node:stream";

type Options = {
  size: number;
};

export function createBatchTransformStream(opts: Options): Transform {
  let currentBatch: unknown[] = [];

  return new Transform({
    objectMode: true,
    transform(chunk, _encoding, callback) {
      currentBatch.push(chunk);

      if (currentBatch.length >= opts.size) {
        this.push(currentBatch);
        currentBatch = [];
      }

      callback();
    },
    flush(callback) {
      if (currentBatch.length > 0) {
        this.push(currentBatch);
      }
      callback();
    },
  });
}
