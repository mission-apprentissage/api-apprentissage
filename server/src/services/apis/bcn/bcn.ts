import { ReadStream } from "node:fs";

import { internal } from "@hapi/boom";
import { isAxiosError } from "axios";
import { ISourceBcn } from "shared/models/source/bcn/source.bcn.model";

import getApiClient from "@/services/apis/client.js";
import { withCause } from "@/services/errors/withCause.js";
import { downloadFileAsStream } from "@/utils/apiUtils.js";

const bcnClient = getApiClient(
  {
    baseURL: "https://infocentre.pleiade.education.fr/bcn",
    timeout: 90_000,
  },
  { cache: false }
);

export async function fetchBcnData(table: ISourceBcn["source"]): Promise<ReadStream> {
  try {
    const response = await bcnClient({
      method: "POST",
      url: "/index.php/export/CSV",
      params: {
        n: table,
        separator: ";",
        withForeign: true,
      },
      responseType: "stream",
    });

    return await downloadFileAsStream(response.data, `bcn_${table}.zip`);
  } catch (error) {
    if (isAxiosError(error)) {
      throw internal("api.bcn: unable to fetchBcnData", { data: error.toJSON() });
    }
    throw withCause(internal("api.bcn: unable to fetchBcnData"), error);
  }
}
