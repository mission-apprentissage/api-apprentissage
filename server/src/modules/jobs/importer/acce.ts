import type { ReadStream } from "node:fs";
import { Transform } from "node:stream";
import { pipeline } from "node:stream/promises";

import { internal } from "@hapi/boom";
import { parse } from "csv-parse";
import { ObjectId } from "mongodb";
import { IAcce, ZAcceByType } from "shared/models/acce/acce.model";
import unzipper from "unzipper";

import parentLogger from "@/common/logger";

import { downloadCsvExtraction } from "../../../common/apis/acce";
import { withCause } from "../../../common/errors/withCause";
import { getDbCollection } from "../../../common/utils/mongodbUtils";
import { createBatchTransformStream } from "../../../common/utils/streamUtils";

const logger = parentLogger.child({ module: "import:acce" });

async function parseAcceFile(stream: ReadStream, source: string, date: Date) {
  try {
    const zod = ZAcceByType[source as keyof typeof ZAcceByType] ?? null;

    if (zod === null) {
      throw internal("import.acce: unexpected source file", { source });
    }

    await pipeline(
      stream,
      parse({
        bom: true,
        columns: true,
        relax_column_count: true,
        encoding: "latin1",
        delimiter: ";",
        trim: true,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        onRecord: (record, { columns }: any) => {
          const data = columns.reduce((acc: Record<string, string | null>, column: { name: string }) => {
            acc[column.name] = record[column.name]?.trim() || null;
            return acc;
          }, {});

          return zod.parse({
            _id: new ObjectId(),
            source,
            date,
            data,
          });
        },
      }),
      createBatchTransformStream({ size: 1_000 }),
      new Transform({
        objectMode: true,
        async transform(chunk, _encoding, callback) {
          try {
            await getDbCollection("source.acce").insertMany(chunk);
            callback();
          } catch (error) {
            callback(withCause(internal("import.acce: error when inserting"), error));
          }
        },
      })
    );

    await getDbCollection("source.acce").deleteMany({
      source: source as IAcce["source"],
      date: { $ne: date },
    });
  } catch (error) {
    throw withCause(internal("import.acce: unable to parseAcceFile", { source }), error);
  }
}

export async function importAcceData(readStream: ReadStream) {
  const importDate = new Date();

  const zip = readStream.pipe(unzipper.Parse({ forceStream: true }));
  for await (const entry of zip) {
    await parseAcceFile(entry, entry.path, importDate);
    entry.autodrain();
  }
}

export async function runAcceImporter() {
  logger.info("Geting ACCE file...");

  const stream = await downloadCsvExtraction();

  logger.info("Import ACCE data starting...");

  await importAcceData(stream);
}
