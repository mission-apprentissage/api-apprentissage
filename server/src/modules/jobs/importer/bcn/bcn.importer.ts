import { Transform } from "node:stream";
import { pipeline } from "node:stream/promises";

import { internal } from "@hapi/boom";
import { parse } from "csv-parse";
import { ObjectId } from "mongodb";
import { ISourceBcn, zBcnBySource } from "shared/models/source/bcn/source.bcn.model";

import { fetchBcnData } from "@/common/apis/bcn/bcn";
import { withCause } from "@/common/errors/withCause";
import parentLogger from "@/common/logger";
import { getDbCollection } from "@/common/utils/mongodbUtils";
import { createBatchTransformStream } from "@/common/utils/streamUtils";

const logger = parentLogger.child({ module: "import:bcn" });

async function importBcnSource(source: ISourceBcn["source"], date: Date): Promise<number> {
  logger.info({ source }, "fetching BCN data");

  try {
    const zod = zBcnBySource[source as keyof typeof zBcnBySource] ?? null;

    if (zod === null) {
      throw internal("import.bcn: unexpected source", { source });
    }

    const stream = await fetchBcnData(source);

    logger.info({ source }, "parsing BCN data");

    await pipeline(
      stream,
      parse({
        bom: true,
        columns: true,
        relax_column_count: true,
        encoding: "latin1",
        delimiter: ";",
        trim: true,
        quote: false,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        onRecord: (record, { columns }: any) => {
          const data = columns.reduce((acc: Record<string, string | null>, column: { name: string }) => {
            // Replace all mongodb dot special character with underscore
            acc[column.name.replaceAll(".", "_")] = record[column.name]?.trim() || null;
            return acc;
          }, {});

          return zod.parse({
            _id: new ObjectId(),
            source,
            date,
            data,
          });
        },
      }),
      createBatchTransformStream({ size: 100 }),
      new Transform({
        objectMode: true,
        async transform(chunk, _encoding, callback) {
          try {
            await getDbCollection("source.bcn").insertMany(chunk);
            callback();
          } catch (error) {
            callback(withCause(internal("import.bcn: error when inserting"), error));
          }
        },
      })
    );

    await getDbCollection("source.bcn").deleteMany({
      source,
      date: { $ne: date },
    });

    return await getDbCollection("source.bcn").countDocuments({ date, source });
  } catch (error) {
    await getDbCollection("source.bcn").deleteMany({ date });
    throw withCause(internal("import.bcn: unable to importBcnSource", { source }), error);
  }
}

export async function runBcnImporter(): Promise<Record<string, number>> {
  const importDate = new Date();

  try {
    const statsBySource: Record<string, number> = {};

    statsBySource["N_FORMATION_DIPLOME"] = await importBcnSource("N_FORMATION_DIPLOME", importDate);
    statsBySource["N_FORMATION_DIPLOME_ENQUETE_51"] = await importBcnSource(
      "N_FORMATION_DIPLOME_ENQUETE_51",
      importDate
    );
    statsBySource["N_NIVEAU_FORMATION_DIPLOME"] = await importBcnSource("N_NIVEAU_FORMATION_DIPLOME", importDate);
    statsBySource["V_FORMATION_DIPLOME"] = await importBcnSource("V_FORMATION_DIPLOME", importDate);

    await getDbCollection("import.meta").insertOne({
      _id: new ObjectId(),
      import_date: importDate,
      type: "bcn",
    });

    return statsBySource;
  } catch (error) {
    throw withCause(internal("import.bcn: unable to runBcnImporter"), error);
  }
}
