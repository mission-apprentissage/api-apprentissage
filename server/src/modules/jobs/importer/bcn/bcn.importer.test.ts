import { afterEach } from "node:test";

import { useMongo } from "@tests/utils/mongo.utils";
import { createReadStream } from "fs";
import { dirname, join } from "path";
import { ISourceBcn } from "shared/models/source/bcn/source.bcn.model";
import { fileURLToPath } from "url";
import { afterAll, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";

import { fetchBcnData } from "../../../../common/apis/bcn/bcn";
import { getDbCollection } from "../../../../common/utils/mongodbUtils";
import { runBcnImporter } from "./bcn.importer";

const mongo = useMongo();

vi.mock("../../../../common/apis/bcn/bcn", async (importOriginal) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const mod = (await importOriginal()) as any;
  return {
    ...mod,
    fetchBcnData: vi.fn(),
  };
});

describe("runBcnImporter", () => {
  beforeAll(async () => {
    await mongo.beforeAll();
  });

  beforeEach(async () => {
    await mongo.beforeEach();
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.mocked(fetchBcnData).mockReset();
  });

  afterAll(async () => {
    vi.useRealTimers();
    await mongo.afterAll();
  });

  it("should import Bcn sources", async () => {
    const date = new Date("2023-04-08T22:00:00.000Z");
    vi.setSystemTime(date);

    vi.mocked(fetchBcnData).mockImplementation(async (source: ISourceBcn["source"]) => {
      const dataFixture = join(dirname(fileURLToPath(import.meta.url)), `fixtures/sample/${source}.csv`);
      return createReadStream(dataFixture);
    });

    await runBcnImporter();

    const coll = getDbCollection("source.bcn");
    const data = await coll.find({}).toArray();

    expect(data.length).toBe(36);
    expect(
      data.map((datum) => ({ ...datum, _id: "ObjectId" })).find((datum) => datum.source === "N_FORMATION_DIPLOME")
    ).toMatchSnapshot();
    expect(
      data
        .map((datum) => ({ ...datum, _id: "ObjectId" }))
        .find((datum) => datum.source === "N_FORMATION_DIPLOME_ENQUETE_51")
    ).toMatchSnapshot();
    expect(
      data
        .map((datum) => ({ ...datum, _id: "ObjectId" }))
        .find((datum) => datum.source === "N_NIVEAU_FORMATION_DIPLOME")
    ).toMatchSnapshot();
    expect(
      data.map((datum) => ({ ...datum, _id: "ObjectId" })).find((datum) => datum.source === "V_FORMATION_DIPLOME")
    ).toMatchSnapshot();

    expect(fetchBcnData).toHaveBeenCalledTimes(4);
    expect(fetchBcnData).toHaveBeenNthCalledWith(1, "N_FORMATION_DIPLOME");
    expect(fetchBcnData).toHaveBeenNthCalledWith(2, "N_FORMATION_DIPLOME_ENQUETE_51");
    expect(fetchBcnData).toHaveBeenNthCalledWith(3, "N_NIVEAU_FORMATION_DIPLOME");
    expect(fetchBcnData).toHaveBeenNthCalledWith(4, "V_FORMATION_DIPLOME");
  });

  it("should throw an error if importBcnSource fails", async () => {
    const date = new Date("2023-04-18T22:00:00.000Z");
    vi.setSystemTime(date);
    const error = new Error("Unable to fetch data");

    vi.mocked(fetchBcnData).mockRejectedValue(error);

    await expect(runBcnImporter()).rejects.toThrowError("import.bcn: unable to runBcnImporter");
    expect(fetchBcnData).toHaveBeenCalledTimes(1);
    expect(fetchBcnData).toHaveBeenNthCalledWith(1, "N_FORMATION_DIPLOME");
  });
});
