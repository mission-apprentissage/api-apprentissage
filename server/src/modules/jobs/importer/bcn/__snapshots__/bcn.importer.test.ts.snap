// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runBcnImporter > should import Bcn sources 1`] = `
{
  "_id": "ObjectId",
  "data": {
    "ANCIEN_DIPLOME_1": "50022304",
    "ANCIEN_DIPLOME_2": null,
    "ANCIEN_DIPLOME_3": null,
    "ANCIEN_DIPLOME_4": null,
    "ANCIEN_DIPLOME_5": null,
    "ANCIEN_DIPLOME_6": null,
    "ANCIEN_DIPLOME_7": null,
    "ANCIEN_RECME": "500835",
    "CITE_DOMAINE_DETAILLE": null,
    "CITE_DOMAINE_FORMATION": "521",
    "COMMENTAIRE": "A REMPLACE LE CAP FABRICATIONS EN ALLIAGES MOULES SUR MODELES (50022304S). ABROGE SANS REMPLACEMENT.",
    "DATE_ARRETE_ABROGATION": "20/03/2007",
    "DATE_ARRETE_CREATION": "06/08/1991",
    "DATE_ARRETE_MODIFICATION_1": null,
    "DATE_ARRETE_MODIFICATION_10": null,
    "DATE_ARRETE_MODIFICATION_2": null,
    "DATE_ARRETE_MODIFICATION_3": null,
    "DATE_ARRETE_MODIFICATION_4": null,
    "DATE_ARRETE_MODIFICATION_5": null,
    "DATE_ARRETE_MODIFICATION_6": null,
    "DATE_ARRETE_MODIFICATION_7": null,
    "DATE_ARRETE_MODIFICATION_8": null,
    "DATE_ARRETE_MODIFICATION_9": null,
    "DATE_DERNIERE_SESSION": "2008",
    "DATE_FERMETURE": "31/08/2009",
    "DATE_INTERVENTION": "10/12/2020",
    "DATE_OUVERTURE": "06/08/1991",
    "DATE_PREMIERE_SESSION": "1993",
    "DATE_SESSION_RATTRAPAGE": "2009",
    "FORMATION_DIPLOME": "50022326",
    "GESTIONNAIRE_FORMATION_DIPLOME": "DGESCO A2-3",
    "GROUPE_SPECIALITE": "223",
    "ID_DOCUMENT": null,
    "LETTRE_SPECIALITE": "S",
    "LIBELLE_COURT": "CAP",
    "LIBELLE_LONG_200": "ALLIAGES MOULES SUR MODELES (CAP)",
    "LIBELLE_STAT_33": "ALLIAGES MOULES SUR MODELES",
    "NATURE_FORMATION_DIPLOME": "1",
    "NB_MEF_FERME": "6",
    "NB_MEF_OUVERT": "0",
    "NIVEAU_FORMATION_DIPLOME": "500",
    "NIVEAU_QUALIFICATION_RNCP": "03",
    "NOUVEAU_DIPLOME_1": null,
    "NOUVEAU_DIPLOME_2": null,
    "NOUVEAU_DIPLOME_3": null,
    "NOUVEAU_DIPLOME_4": null,
    "NOUVEAU_DIPLOME_5": null,
    "NOUVEAU_DIPLOME_6": null,
    "NOUVEAU_DIPLOME_7": null,
    "N_CITE_2013_DOMAINE_DETAILL_LIBELLE_EDITION": null,
    "N_CITE_DOMAINE_FORMATION_LIBELLE_EDITION": "Mécanique et travail du métal",
    "N_COMMENTAIRE": null,
    "N_GROUPE_SPECIALITE_LIBELLE_LONG": "METALLURGIE (Y.C. SIDERURGIE, FONDERIE)",
    "N_LETTRE_SPECIALITE_LIBELLE_LONG": "PRODUCTION",
    "N_NATURE_FORMATION_DIPLOME_LIBELLE_100": "DIPLOME NATIONAL / DIPLOME D'ETAT",
    "N_NIVEAU_FORMATION_DIPLOME_LIBELLE_100": "CERTIFICAT D'APTITUDE PROFESSIONNELLE",
    "N_NIVEAU_QUALIFICATION_RNCP_LIBELLE_LONG": "NIVEAU 3",
    "OBSERVATION": "Grille horaire provisoire n°1.",
    "UNITE_CAPITALISABLE": "O",
  },
  "date": 2023-04-08T22:00:00.000Z,
  "source": "N_FORMATION_DIPLOME",
}
`;

exports[`runBcnImporter > should import Bcn sources 2`] = `
{
  "_id": "ObjectId",
  "data": {
    "ANCIEN_RECME": null,
    "CITE_DOMAINE_DETAILLE": null,
    "CITE_DOMAINE_FORMATION": "850",
    "COMMENTAIRE": "fermï¿½ suite ï¿½ Ar 22/02/2001 Controler libellï¿½ fermï¿½ suite ï¿½ Ar 22/02/2001",
    "DATE_ARRETE_ABROGATION": null,
    "DATE_ARRETE_CREATION": null,
    "DATE_ARRETE_MODIFICATION_1": "22/02/2001",
    "DATE_ARRETE_MODIFICATION_10": null,
    "DATE_ARRETE_MODIFICATION_2": null,
    "DATE_ARRETE_MODIFICATION_3": null,
    "DATE_ARRETE_MODIFICATION_4": null,
    "DATE_ARRETE_MODIFICATION_5": null,
    "DATE_ARRETE_MODIFICATION_6": null,
    "DATE_ARRETE_MODIFICATION_7": null,
    "DATE_ARRETE_MODIFICATION_8": null,
    "DATE_ARRETE_MODIFICATION_9": null,
    "DATE_DERNIERE_SESSION": null,
    "DATE_FERMETURE": "31/08/2001",
    "DATE_INTERVENTION": "06/01/2021",
    "DATE_OUVERTURE": "01/09/1998",
    "DATE_PREMIERE_SESSION": null,
    "DIPLOME_SISE": null,
    "FORMATION_DIPLOME": "1703430A",
    "GESTIONNAIRE_FORMATION_DIPLOME": "SPN",
    "GROUPE_SPECIALITE": "343",
    "HABILITATION_ORIGINE": null,
    "ID_DOCUMENT": null,
    "LETTRE_SPECIALITE": "M",
    "LIBELLE_COURT": "ING",
    "LIBELLE_LONG_200": "INGENIEUR EN HYGIENE INDUSTRIELLE ET ENVIRONNEMENT, DIPLOME DU CONSERVATOIRE NATIONAL DES ARTS ET METIERS CNAM (ING)",
    "LIBELLE_STAT_33": "EN HYGIENE INDUSTRIEL ET ENVIRONM",
    "NATURE_FORMATION_DIPLOME": "1",
    "NIVEAU_FORMATION_DIPLOME": "170",
    "NIVEAU_QUALIFICATION_RNCP": "07",
    "NUMERO_UAI": "0753471R",
    "N_CITE_2013_DOMAINE_DETAILL_LIBELLE_EDITION": null,
    "N_CITE_DOMAINE_FORMATION_LIBELLE_EDITION": "Protection de l'environnement",
    "N_COMMENTAIRE": null,
    "N_DIPLOME_SISE_LIBELLE_INTITULE_1": null,
    "N_GROUPE_SPECIALITE_LIBELLE_LONG": "NETTOYAGE, ASSAIN., PROTECTION ENVIRONMT",
    "N_LETTRE_SPECIALITE_LIBELLE_LONG": "AUTRE/NON INDIQUE (DOMAINE TECHNICO-PRO)",
    "N_NATURE_FORMATION_DIPLOME_LIBELLE_100": "DIPLOME NATIONAL / DIPLOME D'ETAT",
    "N_NIVEAU_FORMATION_DIPLOME_LIBELLE_100": "INGENIEURS RECONNUS (RCT..), NFI , ALTERNANCE",
    "N_NIVEAU_QUALIFICATION_RNCP_LIBELLE_LONG": "NIVEAU 7",
    "UNITE_CAPITALISABLE": null,
  },
  "date": 2023-04-08T22:00:00.000Z,
  "source": "N_FORMATION_DIPLOME_ENQUETE_51",
}
`;

exports[`runBcnImporter > should import Bcn sources 3`] = `
{
  "_id": "ObjectId",
  "data": {
    "ANCIEN_NIVEAU": null,
    "DATE_FERMETURE": null,
    "DATE_INTERVENTION": "10/12/2020",
    "DATE_OUVERTURE": "01/08/1996",
    "LIBELLE_100": "TITRE DE NIV 4  NON HOMOLOGUE - CQP....",
    "LIBELLE_COURT": "TNH4",
    "NIVEAU_FORMATION_DIPLOME": "4ZZ",
    "NIVEAU_INTERMINISTERIEL": "4",
    "NIVEAU_QUALIFICATION_RNCP": "04",
    "N_COMMENTAIRE": null,
    "N_NIVEAU_INTERMINISTERIEL_LIBELLE_LONG": "NIVEAU IV",
    "N_NIVEAU_QUALIFICATION_RNCP_LIBELLE_LONG": "NIVEAU 4",
  },
  "date": 2023-04-08T22:00:00.000Z,
  "source": "N_NIVEAU_FORMATION_DIPLOME",
}
`;

exports[`runBcnImporter > should import Bcn sources 4`] = `
{
  "_id": "ObjectId",
  "data": {
    "ANCIEN_RECME": null,
    "CITE_DOMAINE_FORMATION": "343",
    "COMMENTAIRE": "sup2000 remplace les 13531407 et 13531359",
    "DATE_ARRETE_ABROGATION": null,
    "DATE_ARRETE_CREATION": null,
    "DATE_ARRETE_MODIFICATION_1": null,
    "DATE_ARRETE_MODIFICATION_10": null,
    "DATE_ARRETE_MODIFICATION_2": null,
    "DATE_ARRETE_MODIFICATION_3": null,
    "DATE_ARRETE_MODIFICATION_4": null,
    "DATE_ARRETE_MODIFICATION_5": null,
    "DATE_ARRETE_MODIFICATION_6": null,
    "DATE_ARRETE_MODIFICATION_7": null,
    "DATE_ARRETE_MODIFICATION_8": null,
    "DATE_ARRETE_MODIFICATION_9": null,
    "DATE_DERNIERE_SESSION": null,
    "DATE_FERMETURE": "31/08/2020",
    "DATE_INTERVENTION": "10/12/2020",
    "DATE_OUVERTURE": "01/09/2014",
    "DATE_PREMIERE_SESSION": null,
    "FORMATION_DIPLOME": "13531364",
    "GESTIONNAIRE_FORMATION_DIPLOME": "DEPPA1",
    "GROUPE_SPECIALITE": "313",
    "ID_DOCUMENT": null,
    "LETTRE_SPECIALITE": "M",
    "LIBELLE_COURT": "MASTER PRO",
    "LIBELLE_LONG_200": "FINANCE (MASTER PARIS 12)",
    "LIBELLE_STAT_33": "FINANCE",
    "NATURE_FORMATION_DIPLOME": "1",
    "NIVEAU_FORMATION_DIPLOME": "135",
    "NIVEAU_QUALIFICATION_RNCP": "07",
    "N_CITE_DOMAINE_FORMATION_LIBELLE_EDITION": "Finance, banque, assurance",
    "N_GROUPE_SPECIALITE_LIBELLE_LONG": "FINANCES, BANQUE, ASSURANCES",
    "N_LETTRE_SPECIALITE_LIBELLE_LONG": "AUTRE/NON INDIQUE (DOMAINE TECHNICO-PRO)",
    "N_NATURE_FORMATION_DIPLOME_LIBELLE_100": "DIPLOME NATIONAL / DIPLOME D'ETAT",
    "N_NIVEAU_FORMATION_DIPLOME_LIBELLE_100": "MASTER PROFESSIONNEL",
    "N_NIVEAU_QUALIFICATION_RNCP_LIBELLE_LONG": "NIVEAU 7",
    "UNITE_CAPITALISABLE": null,
  },
  "date": 2023-04-08T22:00:00.000Z,
  "source": "V_FORMATION_DIPLOME",
}
`;
